src/shrubheight/treatment/shrub_stats_sfm.py

```diff
@@ -52,70 +54,121 @@ def compute_stats(dname: str, data: np.ndarray) -> dict:
     return stats_result
 
 
-def get_raster_stats(polygon: NamedTuple, raster_files: list) -> dict:
+def get_raster_stats(polygons: gpd.GeoDataFrame, raster_files: dict) -> dict:
     """Extract and compute statistics from raster data for a given polygon.
 
     Args:
-        polygon: a NamedTuple - GeoDataFrame row with itertuples() - containing polygon geometry
-        raster_files: List of paths to raster files
+        polygons: a GeoDataFrame containing polygon geometry
+        raster_files: dict of raster file data
+            name : { bounds : (tuple)
+                     path: str }
 
     Returns:
         Dictionary of computed statistics
     """
-    stats_d = {}
+    stats = []
 
-    for raster_file in raster_files:
+    for filename in raster_files:
         # TODO - depending on file naming conventions brittle
-        dname = raster_file.split("/")[-1].split("_")[-2]
-        with rasterio.open(raster_file) as src:
-            # If the polygon intersects the bounding box of the raster
-            if not rasterio.coords.disjoint_bounds(src.bounds, polygon.geometry.bounds):
-                # Read the raster data that overlaps with the polygon
-                out_image, _ = mask(src, [polygon.geometry], crop=True)
-
-                if src.count >= 3:
-                    red, green, blue = out_image[:3].astype(float)
-                    data = (green - red) / (green + red - blue + 1e-6)
-                    data = data[data <= 1]
-                else:
-                    no_data = src.nodata
-                    data = out_image[out_image != no_data]
-
-                # Store the results
-                if (data.size > 0) & (data.mean() != 0):
-                    stats_d = compute_stats(dname, data)
-                    # break
-    return stats_d
-
-
-def process_data(input_dir: str, method: str, output_dir: str) -> None:
+        # It's better if we pass in a dict here - more flexible later too
+        dname = file_shortname(filename)
+        with rasterio.open(filename) as src:
+
+            for polygon in tqdm(polygons.itertuples()):
+                stats_d = {"id": int(polygon.id)}
+
+                # If the polygon intersects the bounding box of the raster
+                if not rasterio.coords.disjoint_bounds(
+                    src.bounds, polygon.geometry.bounds
+                ):
+                    # Read the raster data that overlaps with the polygon
+                    data = masked_polygon(src, polygon)
+                    if (data.size > 0) & (data.mean() != 0):

```

Comment:
I'm not sure about the check that `data.mean()` - assuming it's a proxy for checking that all elements are non-zero? Might be worth adding to the docstring that empty/zero-valued polygons are passed over?

Reply:
Aye, I don't know how likely in this scenario a large nodata area in the raster would be.

It's a good hint to update the docstring which had drifted out of date anyway